examples/step-79/step-79.cc

```diff
@@ -0,0 +1,2756 @@
+            else
+              {
+                cells_to_check = cells_to_check_temp;
+                n_neighbors    = neighbor_ids.size();
+              }
+          }
+      }
+
+    // The final step is to normalize the matrix so that for each
+    // row, the sum of entries equals one.
+    for (unsigned int i = 0; i < filter_matrix.m(); ++i)
+      {
+        double denominator = 0;
+        for (SparseMatrix<double>::iterator iter = filter_matrix.begin(i);
+             iter != filter_matrix.end(i);
+             iter++)
+          denominator = denominator + iter->value();
+        for (SparseMatrix<double>::iterator iter = filter_matrix.begin(i);
+             iter != filter_matrix.end(i);
+             iter++)
+          iter->value() = iter->value() / denominator;
+      }
+  }
+
+
+  // @sect3{Assembling the Newton matrix}
+
+  // Whereas the previous function built a matrix that is the same as
+  // long as the mesh does not change (which we don't do anyway in
+  // this program), the next function builds the matrix to be solved
+  // in each iteration. This is where the magic happens. The components
+  // of the system of linear equations describing Newtons' method for
+  // finding the solution of the KKT conditions are implemented here.
+  //
+  // The top of the function is as in most of these functions and just
+  // sets up all sorts of variables necessary for the actual assembly,
+  // including a whole bunch of extractors. The entire set up should
+  // look familiar, though somewhat lengthier, if you've previously
+  // looked at step-22.
+  template <int dim>
+  void SANDTopOpt<dim>::assemble_system()
+  {
+    TimerOutput::Scope t(timer, "assembly");
+
+    system_matrix = 0;
+    system_rhs    = 0;
+
+    const FEValuesExtractors::Scalar densities(
+      SolutionComponents::density<dim>);
+    const FEValuesExtractors::Vector displacements(
+      SolutionComponents::displacement<dim>);
+    const FEValuesExtractors::Scalar unfiltered_densities(
+      SolutionComponents::unfiltered_density<dim>);
+    const FEValuesExtractors::Vector displacement_multipliers(
+      SolutionComponents::displacement_multiplier<dim>);
+    const FEValuesExtractors::Scalar unfiltered_density_multipliers(
+      SolutionComponents::unfiltered_density_multiplier<dim>);
+    const FEValuesExtractors::Scalar density_lower_slacks(
+      SolutionComponents::density_lower_slack<dim>);
+    const FEValuesExtractors::Scalar density_lower_slack_multipliers(
+      SolutionComponents::density_lower_slack_multiplier<dim>);
+    const FEValuesExtractors::Scalar density_upper_slacks(
+      SolutionComponents::density_upper_slack<dim>);
+    const FEValuesExtractors::Scalar density_upper_slack_multipliers(
+      SolutionComponents::density_upper_slack_multiplier<dim>);
+
+    QGauss<dim>       quadrature_formula(fe.degree + 1);
+    QGauss<dim - 1>   face_quadrature_formula(fe.degree + 1);
+    FEValues<dim>     fe_values(fe,
```

Comment:
Could we introduce a mapping?

Reply:
I don't really understand what you're asking - sorry. I took a brief look at the mapping class documentation, but I'm not sure if that's what you're talking about. What would the mapping be for?